### ISSUE TITLE
Support using tekton to create pipelines

### ISSUE THREAD

**Issue by flamywhale** [kubesphere-sigs/ks-devops#81]
Add support for using tekton to create pipelines in `ks-devops`.

/label ospp-2021

**Comment by flamywhale** [comment]
API design is [here](https://docs.google.com/document/d/1nMzb78ArsLRv6z0i-97a1ah3DPxJnjEn91ksDfGnA7Y/edit?usp=sharing).
@shihaoH @LinuxSuRen

**Comment by flamywhale** [comment]
CLI basic design can be seen [here](https://gist.github.com/flamywhale/23c557a3bbcb987c804dc53be5cfcb9b).

**Comment by flamywhale** [comment]
To simplify the usage of `Tekton` in ks-devops, I decide to create a new syntax for using devops referring to [Github Actions](https://docs.github.com/en/actions).

## TODO

What I am going to do is as follows:
1. create a new crd named **action** (pipeline crd has already been used by jenkins).
2. create the controller of the crd **action** to translate it to tekton components and run the tekton-based pipeline.

## Syntax design

Here, I will describe the example of the new devops syntax first to give you a blueprint of it.

### Overview 

* A pipeline is composed of several tasks, and each task consists of plenty of steps.
* A pipeline can specify the executing order of tasks.

```bash
┌─────────────────────────────────┐   ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┐
│         Pipeline                │                          
│  ┌─────────────────────────┐    │   │name: "pipeline name"│
│  │      Task               │    │                          
│  │ ┌─────────────────────┐ │    │   │tasks:               │
│  │ │    Step             │ │    │                          
│  │ └─────────────────────┘ │──┐ │   │  - name: "task1"    │
│  │ ┌─────────────────────┐ │  │ │        steps:            
│  │ │    Step             │ │  │ │   │      - step1        │
│  │ └─────────────────────┘ │  │ │          - step2         
│  └─────────────────────────┘  │ │   │                     │
│  ┌─────────────────────────┐  │ │                          
│  │      Task               │  │ │   │                     │
│  │ ┌─────────────────────┐ │  │ │      - name: "task2"     
│  │ │    Step             │ │  │ │   │    need:            │
│  │ └─────────────────────┘ │◀─┘ │          - task1         
│  │ ┌─────────────────────┐ │    │   │    steps:           │
│  │ │    Step             │ │    │          - step1         
│  │ └─────────────────────┘ │    │   │      - step2        │
│  └─────────────────────────┘    │                          
└─────────────────────────────────┘   └ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘                              
```

### Details of syntax

The detailed syntax can be describe by the following yaml file:
```yaml
name: "pipeline name" # required, specify the pipeline name
tasks: # tasks contains a lot of task
  - name: "task name 1" # required, specify the task name
    need: [""] # names of tasks which must be done before the current task
    uses: "template name of task" # optional, the name of template task
    with: # optional, params passed to the template task
      - name: "param key"
        value: "param value"
    env:
      - name: "environment variable name"
        value: "environment variable value"
    steps:
      - name: "step name 1" # required, the name of the step
        image: "gcr.io/image:tag" # optional, the image url of the container
        workspace: "/path/to/execute/scripts" # optional, set the workspace of the step
        command: ["echo"] # optional, entry point of the container image
        args: ["hello", "step 1"] # optional, arguments of the entry point
        script: | # optional, command line scripts
          # command line scripts
          # e.g.
          ls
          echo "hello step 1"
```

### Use cases

#### hello world
 
The following pipeline will print `hello world`.

```yaml
name: "Hello world"
tasks:
  - name: "echo-hello-world-task"
    steps:
      - name: "echo-hello-world-step"
        image: "ubuntu:16.04"
        command: ["/bin/bash"]
        args: ["-c", "echo hello world"]
```

#### Using git task template

This example is about how to use predefined task template.

```yaml
name: "Git clone"
tasks:
  - name: "git-clone-task"
    uses: "git-cli"
    with:
      - name: "GIT_URL"
        value: "https://github.com/kubesphere/ks-devops"
      - name: "GIT_REVISION"
        value: "master"
      - name: "GIT_SECRET"
        value: "${GIT_TOKEN}"
    env:
      - name: "GIT_TOKEN"
        value: "token"
```
-----

Depending on the above examples, we can briefly describe the usage of an image-based task and a template-based task.
* We can choose only one format of task to work in each task.
  * If we decide to use an image-based task, the keyword `steps` and its content are needed, just like the example `hello world`.
  * If we decide to use a template-based task, the keyword `uses` and `with` are needed. (See the example of `git clone`)
    * `uses` provides the template task name.
    * `with` provides the params needed to pass to the template.

**Comment by flamywhale** [comment]
Based on the above syntax design, we can create a new crd called `action`.

We can define the following structs in `/api/devops/v2alpha1/action_types.go` and using `make manifests` in command line to generate CRD yaml files called `devops.kubesphere.io_actions.yaml`.

```go
type TemplateParam struct {
	Name  string `json:"name" description:"key"`
	Value string `json:"value" description:"value"`
}

type Step struct {
	corev1.Container `json:",inline"`
	Workspace        string `json:"workspace,omitempty" description:"the path to execute step operations"`
}

type TaskSpec struct {
	Name  string          `json:"name" description:"the name of the task"`
	Need  []string        `json:"need,omitempty" description:"tasks which should run before the current task"`
	Steps []Step          `json:"steps,omitempty" description:"steps within the current task"`
	Uses  string          `json:"uses,omitempty" description:"task template name"`
	With  []TemplateParam `json:"with,omitempty" description:"params list of the using template"`
}

// ActionSpec defines the desired state of Action
type ActionSpec struct {
	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
	// Important: Run "make" to regenerate code after modifying this file
	Name  string     `json:"name" description:"the name of the pipeline"`
	Tasks []TaskSpec `json:"tasks,omitempty" description:"tasks defails"`
}

// ActionStatus defines the observed state of Action
type ActionStatus struct {
	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
	// Important: Run "make" to regenerate code after modifying this file
}

// +genclient
// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object

// Action is the Schema for the pipelines API
// +k8s:openapi-gen=true
type Action struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	Spec   ActionSpec   `json:"spec,omitempty"`
	Status ActionStatus `json:"status,omitempty"`
}
```

**Comment by flamywhale** [comment]
@LinuxSuRen Any questions about the above syntax design and the crd implementation are welcome.

**Comment by LinuxSuRen** [comment]
> create a new crd named action (pipeline crd has already been used by Jenkins).

I prefer to use `Pipeline`. It's not a problem because they have a different version.

And one more question, how users can trigger the Pipeline?

**Comment by flamywhale** [comment]
> > create a new crd named action (pipeline crd has already been used by Jenkins).
> 
> I prefer to use `Pipeline`. It's not a problem because they have a different version.

* Now that I get it that `Pipeline` with another version is acceptable in ks-devops. So pipeline is the first choice.

> 
> And one more question, how users can trigger the Pipeline? For example, do we have a CRD names `PipelineRun`?
> 

`PipelineRun` CRD is needed in order to get the running status and the log of the pipeline.

**Comment by flamywhale** [comment]
In addition, [`trigger`](https://github.com/tektoncd/triggers) mechanism is also provided by Tekton. The `trigger` uses the webhook to get messages sent from other git repositories. That is, i.e. we can trigger the git clone repository to run our pipeline. Just like the keyword `on` in [Github Actions](https://docs.github.com/en/actions).

However, [`trigger`](https://github.com/tektoncd/triggers) is another project of Tekton. The [tekton-helm-chart](https://github.com/cdfoundation/tekton-helm-chart) does not include trigger components. If we need to support trigger, we need to install it additionally. This can be done in the future work.

**Comment by LinuxSuRen** [comment]
The word `trigger` has multiple means here.

* User a trigger a Pipeline manually
  * This is what I want to have in this [comment](https://github.com/kubesphere/ks-devops/issues/81#issuecomment-894894287)
* A Pipeline triggered automatically (for example, webhook or APIs)
  * Of course, I agree with @flamywhale , we can support it later. It can be optional.

**Comment by LinuxSuRen** [comment]
Please don't forget the following potential features of the `PipelineRun`

* remove old (or not used history records), aka PipelineRun GC
  * Please consider providing an option for it. For example, limitation of counts, days

**Comment by flamywhale** [comment]
According to the design described in the above [comment](https://github.com/kubesphere/ks-devops/issues/81#issuecomment-894814633), I would like to explain the details of implementation of creating new CRDs and their controller.

## Pipeline

### CRD Design

* CRD definition is in this [comment](https://github.com/kubesphere/ks-devops/issues/81#issuecomment-894815707).
  * The definition is described in the form of `pipeline_types.go` code, which will help us generate crd yaml file in the `kubebuilder` framework.

### Implementation

* The controller of `Pipeline` CRD is necessary to be implemented for translating our `Pipeline` CRD to Tekton CRDs, e.g. `Task`, `Pipeline`, and `PipelineResource`.
* Our translation logic will be implemented in the `reconcile` function in `pipeline_controller.go`.

## PipelineRun

### CRD Design

```
name: pipeline-name
description: run-for-first-time
```

### Implementation

* We need to realize a controller for `PipelineRun` CRD.
* In this controller, we are supposed to trigger the status of the crd, which is different from the logic of  `Pipeline` controller.
